File: mytower/game/views/renderers/floor_renderer.py
# Copyright (c) 2025 Ryan Osterday. All rights reserved.
# See LICENSE file for details.

from typing import Final

import pygame
from pygame import Surface
from pygame.font import Font

from mytower.game.core.constants import FLOORBOARD_HEIGHT
from mytower.game.core.units import Blocks, Pixels, rect_from_pixels
from mytower.game.models.model_snapshots import FloorSnapshot
from mytower.game.utilities.logger import LoggerProvider, MyTowerLogger


class FloorRenderer:

    def __init__(self, logger_provider: LoggerProvider, font: Font) -> None:
        self._logger: MyTowerLogger = logger_provider.get_logger("FloorRenderer")
        self._font: Font = font

    def calculate_floor_bottom_position(self, floor_number: int) -> Pixels:
        """
        Calculates the bottom position (in pixels) of the given floor number.
        Floors are 1-indexed, so floor 1 starts at position 0.
        """
        return Blocks(floor_number - 1).in_pixels


    # TODO: The lower edge depends on the floor below it, so we need to pass that in (in blocks) instead of floor_num
    def draw(self, surface: Surface, floor: FloorSnapshot) -> None:
        screen_height: Pixels = Pixels(surface.get_height())
        floor_height: Pixels = floor.floor_height.in_pixels

        floor_bottom_z: Pixels = self.calculate_floor_bottom_position(floor.floor_number)
        floor_top_z: Pixels = floor_bottom_z + floor_height
        floor_top_y: Pixels = screen_height - floor_top_z
        left_edge: Pixels = floor.left_edge_block.in_pixels
        floor_width: Pixels = floor.floor_width.in_pixels

        # Draw the main floor rectangle
        pygame.draw.rect(
            surface, floor.floor_color, rect_from_pixels(left_edge, floor_top_y, floor_width, floor_height)
        )

        # Draw the floorboard at the top of the floor
        pygame.draw.rect(
            surface, floor.floorboard_color, rect_from_pixels(left_edge, floor_top_y, floor_width, FLOORBOARD_HEIGHT)
        )

        # Optionally draw the floor number for debugging
        text_surface: Final[Surface] = self._font.render(f"Floor {floor.floor_number}", True, (0, 0, 0))
        surface.blit(text_surface, (int(left_edge) + 5, int(floor_top_y) + 5))

File: web/src/rendering/FloorRenderer.ts
// Copyright (c) 2025 Ryan Osterday. All rights reserved.
// See LICENSE file for details.

/**
 * FloorRenderer - Responsible for drawing floors
 * Matches Python's FloorRenderer
 */

import { Blocks } from '../units/Blocks';
import { FLOORBOARD_HEIGHT, PIXELS_PER_BLOCK } from './constants';
import { CoordinateTransform } from './CoordinateTransform';
import type { ColorGQL, FloorSnapshotGQL } from '../generated/graphql';

export class FloorRenderer {
  private transform: CoordinateTransform;

  constructor(
    private context: CanvasRenderingContext2D,
    canvasHeight: number
  ) {
    this.transform = new CoordinateTransform(canvasHeight);
  }

  /**
   * Calculate the bottom position (in world space) of a floor.
   * Floors are 1-indexed, so floor 1 starts at z=0.
   */
  private calculateFloorBottom(floorNumber: number): Blocks {
    return Blocks.from(floorNumber - 1);
  }

  public drawFloor(floor: FloorSnapshotGQL): void {
    // World space (z): positions in blocks from ground
    const floorBottomZ = this.calculateFloorBottom(floor.floorNumber);
    const floorHeight = Blocks.from(floor.floorHeight);
    const floorTopZ = floorBottomZ.add(floorHeight);

    // Screen space (y): convert to pixels from top of canvas
    const floorTopY = this.transform.worldToScreen(floorTopZ);
    
    // Dimensions in pixels
    const leftX = floor.leftEdgeBlock * PIXELS_PER_BLOCK;
    const width = floor.floorWidth * PIXELS_PER_BLOCK;
    const height = floorHeight.toPixels().value;

    // Draw floor background
    this.drawFloorBackground(leftX, floorTopY.value, width, height, floor.floorColor);
    
    // Draw floorboard line at top
    this.drawFloorboard(leftX, floorTopY.value, width, floor.floorboardColor);
    
    // Draw floor number
    this.drawFloorNumber(floor.floorNumber, leftX, floorTopY.value);
  }

  private drawFloorBackground(
    x: number,
    y: number,
    width: number,
    height: number,
    color: ColorGQL
  ): void {
    this.context.fillStyle = `rgb(${color.red}, ${color.green}, ${color.blue})`;
    this.context.fillRect(x, y, width, height);
  }

  private drawFloorboard(
    x: number,
    y: number,
    width: number,
    color: ColorGQL
  ): void {
    this.context.fillStyle = `rgb(${color.red}, ${color.green}, ${color.blue})`;
    this.context.fillRect(x, y, width, FLOORBOARD_HEIGHT);
  }

  private drawFloorNumber(floorNumber: number, x: number, y: number): void {
    this.context.fillStyle = '#000000';
    this.context.font = '18px Helvetica';
    this.context.textAlign = 'left';
    this.context.textBaseline = 'top';
    this.context.fillText(`Floor ${floorNumber}`, x + 8, y + 12);
  }
}